misc-utils/waitpid.c

```diff
@@ -0,0 +1,226 @@
+/*
+ * waitpid(1) - wait for process termination
+ *
+ * Copyright (C) 2022 Thomas Weißschuh <thomas@t-8ch.de>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it would be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include <sys/epoll.h>
+#include <assert.h>
+#include <stdlib.h>
+#include <errno.h>
+#include <string.h>
+#include <unistd.h>
+#include <stdbool.h>
+#include <getopt.h>
+
+#include "pidfd-utils.h"
+#include "c.h"
+#include "nls.h"
+
+#define EXIT_NOT_IMPLEMENTED 2
+
+static bool verbose = false;
+
+static pid_t *parse_pids(size_t n_strings, char * const *strings)
+{
+	pid_t *pids = calloc(n_strings, sizeof(*pids));
+	if (!pids)
+		return NULL;
+
+	for (size_t i = 0; i < n_strings; i++) {
+		char *endptr;
+		const char *arg = strings[i];
+		long long pid = strtoll(arg, &endptr, 10);
+		if (!endptr || pid <= 0) {
+			warnx("invalid PID: '%s'", arg);
+			goto fail;
+		}
+		pids[i] = pid;
+	}
+
+	return pids;
+
+fail:
+	free(pids);
+	return NULL;
+}
+
+static int *open_pidfds(size_t n_pids, pid_t *pids)
+{
+	int *pidfds = calloc(n_pids, sizeof(*pidfds));
+	if (!pidfds)
+		return NULL;
+
+	for (size_t i = 0; i < n_pids; i++) {
+		pidfds[i] = pidfd_open(pids[i], 0);
+		if (pidfds[i] == -1) {
+			warn("could not open pid %d", pids[i]);
+			for (size_t j = 0; j < i; j++)
+				close(pidfds[j]);
+			goto fail;
+		}
+	}
+
+	return pidfds;
+
+fail:
+	free(pidfds);
+	return NULL;
+}
+
+static int add_listeners(int epll, size_t n_pids, int * const pidfds)
+{
+	for (size_t i = 0; i < n_pids; i++) {
+		struct epoll_event evt = {
+			.events = EPOLLIN,
+			.data = { .u64 = i },
+		};
+		int ret = epoll_ctl(epll, EPOLL_CTL_ADD, pidfds[i], &evt);
+		if (ret)
+			return ret;
+	}
+	return 0;
+}
+
+static int wait_for_exits(int epll, size_t n_pids, pid_t * const pids, int * const pidfds)
+{
+	size_t n_left = n_pids;
+	while (n_left) {
+		struct epoll_event evt;
+		int ret = epoll_wait(epll, &evt, 1, -1);
+		if (ret == -1) {
+			if (errno == EINTR)
+				continue;
+			else
+				return -errno;
+		}
+		if (verbose)
+			fprintf(stderr, "PID %d finished\n", pids[evt.data.u64]);
```

Comment:
Use _() to translate. Would it be better to use `stdout` for non-error messages?